plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeHistoryTable.java

```diff
@@ -158,11 +160,21 @@ public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHand
 
         TrinoFileSystem fileSystem = fileSystemFactory.create(session);
         try {
-            List<CommitInfoEntry> commitInfoEntries = TransactionLogTail.loadNewTail(fileSystem, tableLocation, startVersionExclusive, endVersionInclusive).getFileEntries().stream()
+            ImmutableList.Builder<CommitInfoEntry> commitInfoEntries = ImmutableList.builder();
+            Optional<Long> middleVersion = lastCheckpointVersion;
+            if (lastCheckpointVersion.isPresent() && endVersionInclusive.isPresent()) {
+                middleVersion = Optional.of(Math.min(lastCheckpointVersion.get(), endVersionInclusive.get()));
+            }
+            commitInfoEntries.addAll(TransactionLogTail.loadNewTailBackward(fileSystem, tableLocation, startVersionExclusive, middleVersion, true).getFileEntries().stream()
```

Comment:
Why do you need to do this from the middle backwards and then from the middle forward separately? Rather than from endVersion backwards in one pass.

It's not obvious to me this way that you're not potentially going to have duplicates.